### PR TITLE
[WIP] Make multithreading work with independent tbb::task_arena instead of a tbb::task_scheduler singleton

### PR DESCRIPTION
**Pull request by xvallspl**
***tl;dr;*** This PR eliminates (almost all of) the dependencies between the implicit and explicit multithreading execution modes in ROOT and allows the co-existance of several TThreadExecutor instances, each one executing on a different number of threads. 

It doesn't change any other behaviour in the eyes of the user.

TO DO:

- [ ] Decide on explicit-implicit MT execution modes' interactions
- [ ] Rename TPoolManager
- [ ] Decide if allowing change of the number of threads when calling several times EnableImplicitMT(x) with a varying x without an intermediate call to DisableImplicitMT()
- [ ] Add warnings 

***********************************************************************************

# Previous behaviour
Previous to this PR, the number of threads was limited during the lifetime of the tbb scheduler, kept alive by TPoolManager as a ```std::shared_ptr``` as long as references to it existed, to the value set on its initialization.

```cpp
//We initialize the scheduler with 4 threads
ROOT::EnableIMT(4);
{ 
    //The scheduler is active, so the value passed to TThreadExecutor
    //is overriden with the number of threads the scheduler has been
    //initialized with (4)
    TThreadExecutor executor(9);
}
ROOT::DisableIMT();

//The scheduler is not alive at this point, 
//so we initialize it with 2 threads
ROOT::EnableIMT(2);
ROOT::TThreadExecutor executor(8);
ROOT::DisableIMT();

executor.MapReduce(...); //Runs on two threads!

ROOT::EnableIMT(3);
//Still two threads! TThreadExecutor instance was keeping the scheduler alive
```
This also implies that given two co-existent instances of TThreadExecutor initialized with a different number of threads, the first one to be initialized forces on the second one the number of threads to work with.

```cpp
ROOT::TThreadExecutor executor(4);
ROOT::TThreadExecutor executor2(8); //will run limited to two threads!!
```
This setup was useful [to avoid undefined behaviours between the implicit and explicit parallelism modes of ROOT](https://indico.cern.ch/event/607814/contributions/2466931/attachments/1409778/2155811/TScheduler.pdf). 

# New behaviour
With this PR, we can separate this two modes of execution. We support the co-existance of several TThreadExecutors, each handling a different number of threads, by initializing the scheduler with the [default number of threads](tbb::task_scheduler_init::default_num_threads) and using a ```tbb::task_arena``` per TThreadExecutor to work with a subset of them instead.

```cpp

//Each of the executor manages its own tbb::task_arena,
//which allows the co-existance of TThreadExecutors
//handling different number of threads.
TThreadExecutor executor1(8); //will run on 8 threads
TThreadExecutor executor2(4); //will run on 4 threads

//IMT keeps a different task Arena too!
ROOT::EnableIMT(4); //4 threads will be used in IMT operations
//executor3 will be initialized with 4 threads for backward
//compatibility. Should we not allow this interaction? 
//Should it be initialized with the default number of threads?
TThreadExecutor executor3; //Implicit constructor. Initialized with 4 threads.
ROOT::DisableIMT();

ROOT::EnableIMT(2); //2 threads will be used in IMT operations
ROOT::TThreadExecutor executor(8); //Explicit number of threads. 
                                   // Will execute on 8 threads.
ROOT::EnableIMT(4); //2 threads will be used in IMT operations
                    //Doesn't change until disabled! 
                    //Should we allow it instead?
ROOT::DisableIMT();
```

We still destroy the scheduler when not in use, and TPoolManager only reason to exist is to manage it's lifetime (needs a name change):

```cpp
{
//TThreadExecutor holds a shared_ptr to the tbb::task_scheduler
TThreadExecutor executor1(8); //will run on 8 threads
}

//executor1 went out of scope and was destroyed together with the scheduler.
//No scheduler active at this point.

ROOT::EnableIMT(4);

//"IMT" holds holds a shared_ptr to the tbb::task_scheduler. Scheduler alive here.
// DisableIMT() will destroy the IMT reference to the scheduler. The reference count of
// the scheduler reaches zero and it gets destroyed.
ROOT::DisableIMT();


ROOT::EnableIMT(4);
TThreadExecutor executor3; //Implicit constructor. Initialized with 4 threads.
ROOT::DisableIMT();

//The scheduler is still alive here because of executor3
```

### CODE REVIEW

**Comment by xvallspl** [comment]
@phsft-bot build with -Dimt=OFF

File: core/imt/src/TPoolManager.cxx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#include "ROOT/TPoolManager.hxx"
+#include "TPoolManager.hxx"
```
Comment (amadio) [inline core/imt/src/TPoolManager.cxx]: Isn't this a breaking change for users?

File: core/imt/src/TPoolManager.cxx [hunk 2]
```diff
@@ -35,21 +31,16 @@ namespace ROOT {
...
-
       //Factory function returning a shared pointer to the only instance of the PoolManager.
-      std::shared_ptr<TPoolManager> GetPoolManager(UInt_t nThreads)
+      std::shared_ptr<TPoolManager> GetPoolManager()
       {
          if (GetWP().expired()) {
-            std::shared_ptr<TPoolManager> shared(new TPoolManager(nThreads));
+            // why not make_shared: https://stackoverflow.com/posts/20895705/
+            //(under "Why do instances of weak_ptrs keep the control block alive?"")
+            std::shared_ptr<TPoolManager> shared(new TPoolManager());
```
Comment (amadio) [inline core/imt/src/TPoolManager.cxx]: I get a 404 for the link above. Consider adding the explanation here directly as to why you're not using make_shared, since that's a C++11 thing.

File: core/imt/src/TImplicitMT.cxx [hunk 3]
```diff
@@ -29,6 +30,19 @@ static std::shared_ptr<ROOT::Internal::TPoolManager> &R__GetPoolManagerMT()
    return schedMT;
 }
 
+UInt_t &R__GetIMTPoolSize()
+{
+   static UInt_t poolSize;
+   return poolSize;
+}
+
+static std::shared_ptr<tbb::task_arena> &R__GetIMTTaskArena()
```
Comment (amadio) [inline core/imt/src/TImplicitMT.cxx]: If this is local to the file and static, what's the problem with just declaring a `static tbb::task_arena gIMTArena;` at the top of the file?

File: core/imt/CMakeLists.txt [hunk 4]
```diff
@@ -8,7 +8,7 @@ set(sources base.cxx TTaskGroup.cxx)
 if (imt)
   include_directories(SYSTEM ${TBB_INCLUDE_DIRS})
 
-  set(headers ${headers} ROOT/TPoolManager.hxx ROOT/TThreadExecutor.hxx ROOT/TFuture.hxx)
+  set(headers ${headers} ROOT/TThreadExecutor.hxx ROOT/TFuture.hxx)
```
Comment (amadio) [inline core/imt/CMakeLists.txt]: Has `TPoolManager.hxx` become internal to ROOT? This is just removing it from the dependencies, but I think you meant to not install it, maybe?

**Comment by amadio** [comment]
Hi Xavi, very nice. I have a few comments, but overall I think this is much better than before.

File: core/imt/CMakeLists.txt (hunk 4 above)
Comment (xvallspl) [reply core/imt/CMakeLists.txt]: It was in ROOT::Internal already, but should be in the /src directory instead of the /inc one. I moved it there :)

File: core/imt/src/TPoolManager.cxx (hunk 1 above)
Comment (xvallspl) [reply core/imt/src/TPoolManager.cxx]: No. It's internal to ROOT and I'm the only one that has ever used it afaik

File: core/imt/src/TPoolManager.cxx (hunk 2 above)
Comment (xvallspl) [reply core/imt/src/TPoolManager.cxx]: TPoolManager works as a singleton that is referenced from several parts in ROOT (IMT, TThreadExecutor, TTreeProcessorMT...). We manage it's lifetime with a shared pointer, which provides automatic bookkeeping and allows us to destroy the scheduler when no one is holding a pointer to it. Destroying the scheduler is necessary, for instance, to change the number of threads.

File: core/imt/src/TImplicitMT.cxx (hunk 3 above)
Comment (xvallspl) [reply core/imt/src/TImplicitMT.cxx]: The idea was this not to be static for the file with further upcoming changes... meaning It slipped into the wrong commit. Will remove as it is now.

File: core/imt/src/TPoolManager.cxx (hunk 2 above)
Comment (xvallspl) [reply core/imt/src/TPoolManager.cxx]: I will look about the comment or fix the link, thx

**Comment by xvallspl** [comment]
@phsft-bot build with flags -Dimt=OFF

**Comment by Axel-Naumann** [comment]
@phsft-bot build again, please, now that I fixed a couple of things in jenkins/ctest.

**Comment by xvallspl** [comment]
Closing this until I have an answer for the changes requested in the PPP meeting.
